File: algorithms/java/src/dynamicProgramming/uniquePaths/uniquePathsII.java
// Source : https://oj.leetcode.com/problems/unique-paths-ii/
// Inspired by : http://www.jiuzhang.com/solutions/unique-paths/
// Author : Lei Cao
// Date   : 2015-10-11

/**********************************************************************************
 *
 * Follow up for "Unique Paths":
 *
 * Now consider if some obstacles are added to the grids. How many unique paths would there be?
 *
 * An obstacle and empty space is marked as 1 and 0 respectively in the grid.
 *
 * For example,
 * There is one obstacle in the middle of a 3x3 grid as illustrated below.
 *
 * [
 *   [0,0,0],
 *   [0,1,0],
 *   [0,0,0]
 * ]
 *
 * The total number of unique paths is 2.
 *
 * Note: m and n will be at most 100.
 *
 **********************************************************************************/

package dynamicProgramming.uniquePaths;

public class uniquePathsII {
    /**
     * @param obstacleGrid: A list of lists of integers
     * @return: An integer
     */
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid.length == 0 || obstacleGrid[0].length ==0) {
            return 0;
        }
        if (obstacleGrid[0][0] == 1) {
            return 0;
        }
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        // write your code here
        int[][] matrix = new int[m][n];
        for (int i = 0; i < m; i++) {
            if (obstacleGrid[i][0] != 1) {
                matrix[i][0] = 1;
            } else {
                break;
            }
        }
        for (int i = 0; i < n; i++) {
            if (obstacleGrid[0][i] != 1) {
                matrix[0][i] = 1;
            } else {
                break;
            }
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (obstacleGrid[i][j] == 1) {
                    matrix[i][j] = 0;
                } else {
                    matrix[i][j] = matrix[i-1][j] + matrix[i][j-1];
                }
            }
        }
        return matrix[m-1][n-1];
    }
}

File: algorithms/cpp/uniquePaths/uniquePaths.II.cpp
// Source : https://oj.leetcode.com/problems/unique-paths-ii/
// Author : Hao Chen
// Date   : 2014-06-25

/********************************************************************************** 
* 
* Follow up for "Unique Paths":
* 
* Now consider if some obstacles are added to the grids. How many unique paths would there be?
* 
* An obstacle and empty space is marked as 1 and 0 respectively in the grid.
* 
* For example,
* There is one obstacle in the middle of a 3x3 grid as illustrated below.
* 
* [
*   [0,0,0],
*   [0,1,0],
*   [0,0,0]
* ]
* 
* The total number of unique paths is 2.
* 
* Note: m and n will be at most 100.
*               
**********************************************************************************/

#include <iostream>
#include <vector>
using namespace std;

//As same as DP solution with "Unique Path I", just need to consider the obstacles.
int uniquePathsWithObstacles(vector<vector<int> > &obstacleGrid) {
    vector< vector<int> > v = obstacleGrid;
    unsigned int max=0;
    for (int i=0; i<obstacleGrid.size(); i++){
        for (int j=0; j<obstacleGrid[i].size(); j++){
            if(obstacleGrid[i][j] == 1){
                max = v[i][j] = 0;
            } else {
                if (i>0 && j>0) {
                    max= v[i][j] = v[i-1][j] + v[i][j-1];
                }else if(i>0){
                    max = v[i][j] = v[i-1][j];
                }else if(j>0){
                    max = v[i][j] = v[i][j-1];
                }else{
                    max = v[i][j] = 1 ;
                }
            }
        }
    }
    return max;
}


